src/sc/sidechain.cpp

```diff
@@ -757,61 +769,119 @@ void CSidechain::UpdateScFees(const CScCertificateView& certView)
         __func__, __LINE__, ftScFee, mbtrScFee, scFees.size());
 
-    scFees.push_back(Sidechain::ScFeeData(ftScFee, mbtrScFee));
+    // v1 sidechains
+    if (!isNonCeasing()) {
+        scFees.emplace_back(ftScFee, mbtrScFee);
 
-    // check if we are past the buffer size
-    int delta = scFees.size() - maxSizeOfScFeesContainers;
-    if (delta > 0)
-    {
         // remove from the front as many elements are needed to be within the circular buffer size
         // --
         // usually this is just one element, but in regtest a node can set the max size via a startup option
         // therefore such size might be lesser than scFee size after a node restart
-        while (delta > 0)
+        const size_t max_size { static_cast<size_t>(maxSizeOfScFeesContainers) };
+        while (scFees.size() > max_size)
         {
-            if (scFees.empty())
-                break;
-
             const auto& entry = scFees.front();
-            LogPrint("sc", "%s():%d - popping f=%d, m=%d from list\n",
+            LogPrint("sc", "%s():%d - popping f=%d, m=%d from list, as scFees maxsize has been reached\n",
                 __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee);
             scFees.pop_front();
-            delta--;
+        }
+    }
+    // v2 sidechains
+    else {
+        auto scFeeIt = std::find_if(scFees_v2.begin(), scFees_v2.end(),
+            [&blockHeight](Sidechain::ScFeeData_v2 scFeeElem) { return scFeeElem.submissionHeight == blockHeight; });
+
+        // We have not found a scFeeData for the current height, so we add a new one and perform all the checks
+        // on the container size / element age
+        if (scFeeIt == scFees_v2.end()) {
+            scFees_v2.emplace_back(ftScFee, mbtrScFee, blockHeight);
+
+            // as for v1 sidechains
+            const size_t max_size { static_cast<size_t>(maxSizeOfScFeesContainers) };
+            while (scFees_v2.size() > max_size)
+            {
+                const auto& entry = scFees_v2.front();
+                LogPrint("sc", "%s():%d - popping f=%d, m=%d, h=%d from list, as scFees maxsize has been reached\n",
+                    __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee, entry.submissionHeight);
+                scFees_v2.pop_front();
+            }
+
+            // Also purge all elements submitted within a block that now is too old to be considered
+            scFees_v2.remove_if([blockHeight, this](Sidechain::ScFeeData_v2 entry) {
+                const bool testCondition = (entry.submissionHeight <= blockHeight - this->maxSizeOfScFeesContainers);
+                if (testCondition)
+                    LogPrint("sc", "%s():%d - popping f=%d, m=%d, h=%d from list, as entry is too old\n",
+                        __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee, entry.submissionHeight);
+                return testCondition;
+            });
```

Comment:
Humble suggestion (capture as little as possible in lambdas - especially `this` if not needed)
For all the rest LGTM
Thank you !!!

```suggestion
            // Also purge all elements submitted within a block that now is too old to be considered
            const auto threshold{blockHeight - this->maxSizeOfScFeesContainers};
            scFees_v2.remove_if([threshold](Sidechain::ScFeeData_v2 entry) {
                const bool testCondition = (entry.submissionHeight <= threshold);
                if (testCondition)
                    LogPrint("sc", "%s():%d - popping f=%d, m=%d, h=%d from list, as entry is too old\n",
                        __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee, entry.submissionHeight);
                return testCondition;
            });
```

Reply:
fixed in c3fa17e38bb8677927f9b87bf564d3f8ec9954fc